[DatePicker] New DatePicker component in new gestalt-datepicker package

AlbertCarreras opened the pull request:
This PR creates new component: `DatePicker`.

DatePicker introduces new dependencies: `react-datepicker`. To maintain the lightness of `gestalt`, `DatePicker` is introduced in a new package `gestalt-datepicker`.

To use it, it must be installed and imported: `import Datepicker from "gestalt-datepicker"`

This PR has these major changes:
- refactors `rollup-config.js` to allow multiple gestalt packages.
- hook up to automatic releases
- creates a `gestalt-datepicker` package with the DatePicker component, flow-typed coverage, and tests.
- updates `gestalt/docs` to include a `beta` DatePicker component and multiple use cases.

## TODO

- [x] Accessibility checkup
- [x] Update documentation
- [x] Add/update Tests 
- [x] [Format PR title](https://github.com/pinterest/gestalt/#releasing): `ComponentName: Description`.

AlbertCarreras commented on packages/gestalt-datepicker/src/LocaleDataTypes.js:
```diff
@@ -0,0 +1,146 @@
+// @flow strict
```
need to be fixed

AlbertCarreras commented on packages/gestalt-datepicker/src/DatePicker.js:
```diff
@@ -0,0 +1,193 @@
+  useEffect(() => {
+    if (localeData) {
+      registerLocale(localeData.code, localeData);
+      // $FlowFixMe
```
need to be fixed

jennyscript commented on docs/src/components/Example.js:
```diff
@@ -69,7 +93,8 @@ render(<Example />, document.querySelector("#root"));`,
         content: `import React from "react";
 import "../node_modules/gestalt/dist/gestalt.css";
-import { ${components.join(', ')} } from "gestalt";
+// import "../node_modules/gestalt-datepicker/dist/gestalt-datepicker.css";
```
This can be removed?

AlbertCarreras replied:
I fixed it. I'm converting this logic to functions so they automatically add imports and dependencies based on used packages.

jennyscript commented on docs/src/components/PageHeader.js:
```diff
@@ -16,11 +17,22 @@ const githubUrl = component =>
+            <Tooltip inline text="Do not use in production code">
+              <Badge text="Beta" position="top" />
+            </Tooltip>
+          ) : null}
```
Nice!

jennyscript commented on packages/gestalt-datepicker/.npmignore:
```diff
@@ -0,0 +1,2 @@
+**/__*__/**
```
Just wondering what this path is here to match?

AlbertCarreras replied:
This was on /gestalt
I'm don't know what this is used for. Maybe can be removed

btraut replied:
I read it as files that are in the form of `__FOO__`. Looks like `__snapshots__` fit that description.

AlbertCarreras replied:
it feels to me like it's doing nothing, but not sure

christianvuerings replied:
it was to remain consistent but we no longer have these directories

![image](https://user-images.githubusercontent.com/127199/84550647-02450f00-acc0-11ea-9cb1-15e562ed1b70.png)

jennyscript commented on packages/gestalt-datepicker/dist/gestalt-datepicker.js.flow:
```diff
@@ -0,0 +1,2 @@
+// @flow strict-local
```
Does this need to be strict-local since there aren't any dependencies in here?

AlbertCarreras replied:
Fixed.

jennyscript commented on packages/gestalt-datepicker/src/DatePickerTextField.js:
```diff
@@ -0,0 +1,98 @@
+  isRTL?: boolean,
+|};
+
+type InjectedProps = {|
```
The name InjectedProps sounds like and HOC so it took me a minute to realize these are props that react-datepicker adds on. Could you make the name more obvious or add a comment about that here?

AlbertCarreras replied:
FIXED. thanks for pointing out

christianvuerings commented:
![image](https://user-images.githubusercontent.com/127199/84448254-fd6c5680-abfe-11ea-8310-f46fc9e3257e.png)
Looks like this will break the CodeSandBox integration - can we make sure to include both `gestalt` and `gestalt-datepicker` when clicking on a react datepicker example?

christianvuerings commented:
Could we hook this up to automatic releases? Otherwise, we will never `npm publish gestalt-datepicker`

Looks like the package is still available: https://www.npmjs.com/package/gestalt-datepicker

christianvuerings commented on docs/src/DatePicker.doc.js:
```diff
@@ -0,0 +1,399 @@
+      accessibilityLabelNext='fakeAccessibilityLabelNext'
+      accessibilityLabelPrevious='fakeAccessibilityLabelPrevious'
+      accessibilityLabelIcon='fakeAccessibilityLabelIcon'
+      helperText={'https://pinterest.com/'}
```
@AlbertCarreras could we provide a better helperText? E.g. something like `Use mm/dd/yyyy` as the date format

AlbertCarreras replied:
Fixed

christianvuerings commented:
@AlbertCarreras I noticed that keyboard input (regular typing of the date) does not seem to work. Could you investigate? It does work on https://reactdatepicker.com/

![image](https://user-images.githubusercontent.com/127199/84448713-4c66bb80-ac00-11ea-80fc-c41dfe032546.png)

christianvuerings commented on docs/src/DatePicker.doc.js:
```diff
@@ -0,0 +1,399 @@
+          'Accessibility label for the calendar icon on the date field input that clients such as VoiceOver will read to describe the element.',
+      },
+      {
+        name: 'accessibilityLabelNext',
```
@AlbertCarreras can we rename this to `accessibilityLabelNextMonth`?

AlbertCarreras replied:
fixed

christianvuerings commented on docs/src/DatePicker.doc.js:
```diff
@@ -0,0 +1,399 @@
+          'Accessibility label for the next-month button on the calendar view that clients such as VoiceOver will read to describe the element.',
+      },
+      {
+        name: 'accessibilityLabelPrevious',
```
@AlbertCarreras can we rename this to `accessibilityLabelPreviousMonth`?

AlbertCarreras replied:
fixed

christianvuerings commented on packages/gestalt-datepicker/package.json:
```diff
@@ -0,0 +1,54 @@
+  "version": "2.2.1",
+  "license": "Apache-2.0",
+  "homepage": "https://pinterest.github.io/gestalt",
+  "description": "A set of React UI components which enforce Pinterest’s design language",
```
@AlbertCarreras let's improve this description and make it date picker specific.

AlbertCarreras replied:
fixed

christianvuerings commented on packages/gestalt-datepicker/src/DatePicker.css:
```diff
@@ -0,0 +1,310 @@
+  --borderRadius-halfCircle-End: 0 50% 50% 0;
+  --borderRadius-halfCircle-Start: 50% 0 0 50%;
+  --day-dimensions: 40px;
+  --empty-content: "";
```
@AlbertCarreras let's inline this, not sure why we need a separate variable

AlbertCarreras replied:
fixed

christianvuerings commented on packages/gestalt-datepicker/src/DatePicker.css:
```diff
@@ -0,0 +1,310 @@
+}
+
+.react-datepicker-popper[data-placement^="top"] {
+  /* !important is required to keep same distance from Textfield
```
@AlbertCarreras let's avoid this, could you use a different hack instead?

`.react-datepicker-popper.react-datepicker-popper[data-placement^="top"]` could also be a potential fix (without using `!important`

AlbertCarreras replied:
It does work
![image](https://user-images.githubusercontent.com/10593890/84451602-702dff80-ac08-11ea-803d-31715a38ce15.png)
with !important
![image](https://user-images.githubusercontent.com/10593890/84451655-8b990a80-ac08-11ea-8a22-dc94272002ce.png)
with .react-datepicker-popper.react-datepicker-popper[data-placement^="top"]
![image](https://user-images.githubusercontent.com/10593890/84451792-e894c080-ac08-11ea-89bc-302873e352db.png)

christianvuerings commented on packages/gestalt-datepicker/src/LocaleDataTypes.js:
```diff
@@ -0,0 +1,67 @@
+// flowlint unclear-type:off
+export type LocaleData = {
+  code?: string,
+  formatDistance?: (...args: Array<any>) => any,
```
@AlbertCarreras would there be a way to be more explicit with these types? Can we avoid using `any`

AlbertCarreras replied:
It's the official `date-fns` type.

AlbertCarreras replied:
https://github.com/date-fns/date-fns/blob/master/src/locale/en-AU/index.js.flow

christianvuerings commented on packages/gestalt-datepicker/src/DatePicker.jsdom.test.js:
```diff
@@ -0,0 +1,98 @@
+
+    fireEvent.click(selectedDay);
+
+    expect(mockOnChange).toHaveBeenCalledWith(
```
thanks for adding these great tests!

christianvuerings commented on packages/gestalt-datepicker/src/DatePickerTextField.js:
```diff
@@ -0,0 +1,100 @@
+  } = props;
+
+  return (
+    <Touchable disabled={disabled} onTouch={onClick}>
```
@AlbertCarreras can we avoid wrapping this in a `<Touchable />` - otherwise we will have 2 focus states, one on Touchable and one on TextField:

![image](https://user-images.githubusercontent.com/127199/84450198-4e327e00-ac04-11ea-88a1-d17fe62369af.png)
![image](https://user-images.githubusercontent.com/127199/84450209-55598c00-ac04-11ea-9ba2-4a4d8bde20aa.png)

AlbertCarreras replied:
fixed, comment here https://www.loom.com/share/e4a93927ab0d46f4be67689b8d70becf

AlbertCarreras commented:
> ![image](https://user-images.githubusercontent.com/127199/84448254-fd6c5680-abfe-11ea-8310-f46fc9e3257e.png)
> Looks like this will break the CodeSandBox integration - can we make sure to include both `gestalt` and `gestalt-datepicker` when clicking on a react datepicker example?

How do you know it will break? Atm fails because the dependency 'gestalt-datepicker' = 'latest' doesn't exist yet. 

Do I still both?

AlbertCarreras commented on docs/src/DatePicker.doc.js:
```diff
@@ -0,0 +1,399 @@
+      },
+      { name: 'label', type: 'string' },
+      {
+        name: 'localeData',
```
fixed

christianvuerings replied:
@AlbertCarreras let's follow up and create a code example:

```
import DatePicker from 'gestalt-datepicker';
import {
  cs,
  da
} from 'date-fns/locale';

<DatePicker ...
```

AlbertCarreras replied:
fixed

AlbertCarreras commented:
> @AlbertCarreras I noticed that keyboard input (regular typing of the date) does not seem to work. Could you investigate? It does work on https://reactdatepicker.com/
> 
> ![image](https://user-images.githubusercontent.com/127199/84448713-4c66bb80-ac00-11ea-80fc-c41dfe032546.png)

fixed

AlbertCarreras commented on .github/workflows/release.yml:
```diff
@@ -85,3 +85,7 @@ jobs:
         run: |
           cd packages/gestalt
           yarn publish --registry=https://registry.npmjs.org --no-git-tag-version --new-version ${{ steps.release.outputs.VERSION }}
+          cd packages/gestalt-datepicker
```
I added this here for the npm release. I believe this is the automatic workflow that would publish to npm.

AlbertCarreras commented on scripts/publish.js:
```diff
@@ -14,10 +14,15 @@ const { version } = json;
-  `yarn publish --registry=https://registry.npmjs.org --no-git-tag-version --new-version ${version}`
-);
+// `yarn publish` publishes the package defined by the package.json in the current directory.
+const packages = ['gestalt', 'gestalt-datepicker'];
```
I modified this file but I think the release.yml is the file that executes the publishing currently.

christianvuerings commented on .gitignore:
```diff
@@ -6,6 +6,10 @@ stats.html
 screenshots
 .vscode
 
+packages/gestalt-core/dist/*
```
@AlbertCarreras as a follow up, let's see if we can remove it.

AlbertCarreras replied:
fixed

christianvuerings commented on docs/src/DatePicker.doc.js:
```diff
@@ -0,0 +1,398 @@
+      onChange={({value}) => setDate(value)}
+      value={date}
+    />
+
```
@AlbertCarreras let's follow up on this and remove the extra line break

AlbertCarreras replied:
fixed

christianvuerings commented on docs/src/DatePicker.doc.js:
```diff
@@ -0,0 +1,398 @@
+  const [date, setDate] = React.useState(new Date());
+  return (
+    <DatePicker
+      accessibilityLabelNextMonth='fakeaccessibilityLabelNextMonth'
```
@AlbertCarreras `accessibilityLabelNextMonth="Next month"`

AlbertCarreras replied:
fixed

christianvuerings commented on docs/src/DatePicker.doc.js:
```diff
@@ -0,0 +1,398 @@
+  const handleSetEndDate = ({ event, value }) => setEndDate(value);
+
+  return (
+    <Box>
```
@AlbertCarreras let's remove this extra Box

AlbertCarreras replied:
fixed

christianvuerings commented on docs/src/DatePicker.doc.js:
```diff
@@ -0,0 +1,398 @@
+
+  return (
+    <Box>
+      <Box
```
@AlbertCarreras Can we use `<Row gap={1} justifyContext="between" />`

AlbertCarreras replied:
fixed

christianvuerings commented on docs/src/DatePicker.doc.js:
```diff
@@ -0,0 +1,398 @@
+        <Box column={6} paddingX={1}>
+          <DatePicker
+            startDate={startDate}
+            accessibilityLabelNextMonth='Go to next month'
```
@AlbertCarreras can we replace the single quotes with double quotes in the example

AlbertCarreras replied:
fixed

christianvuerings commented on docs/src/DatePicker.doc.js:
```diff
@@ -0,0 +1,398 @@
+            id='1'
+            idealDirection="up"
+            onChange={handleSetStartDate}
+
```
@AlbertCarreras let's remove the extra line break

AlbertCarreras replied:
fixed

christianvuerings commented on docs/src/DatePicker.doc.js:
```diff
@@ -0,0 +1,398 @@
+        href: 'rangePicker',
+      },
+      {
+        name: 'selectsStart',
```
@AlbertCarreras let's change this to an enum instead: `select: 'start' | 'end'` which will combine `selectsEnd` and `selectsStart`

AlbertCarreras replied:
fixed

christianvuerings commented on docs/src/DatePicker.doc.js:
```diff
@@ -0,0 +1,398 @@
+        description: 'Array of enabled dates for selection',
+      },
+      {
+        name: 'isRTL',
```
@AlbertCarreras can we change this to `layoutDirection: 'rtl' | 'ltr'`?

AlbertCarreras replied:
reminder changer arrow css

AlbertCarreras replied:
fixed

christianvuerings commented on docs/src/DatePicker.doc.js:
```diff
@@ -0,0 +1,398 @@
+      accessibilityLabelNextMonth='fakeaccessibilityLabelNextMonth'
+      accessibilityLabelPreviousMonth='fakeaccessibilityLabelPreviousMonth'
+      accessibilityLabelIcon='fakeAccessibilityLabelIcon'
+      id="example-basic"
```
@AlbertCarreras let's add a label here: `label="Alberto's Birth Date"`

AlbertCarreras replied:
fixed

christianvuerings commented on docs/src/DatePicker.doc.js:
```diff
@@ -0,0 +1,398 @@
+            accessibilityLabelPreviousMonth='Go to previous month'
+            accessibilityLabelIcon='Select a start date'
+            endDate={endDate}
+            id='1'
```
@AlbertCarreras let's change the id's to be something like `start-date` + add a label

AlbertCarreras replied:
fixed

christianvuerings commented on docs/src/DatePicker.doc.js:
```diff
@@ -0,0 +1,398 @@
+      accessibilityLabelPreviousMonth='fakeaccessibilityLabelPreviousMonth'
+      accessibilityLabelIcon='fakeAccessibilityLabelIcon'
+      errorMessage={!date ? "This field can't be blank!" : null}
+      helperText={'https://pinterest.com/'}
```
@AlbertCarreras can we update this to be something else? Or leave it empty

AlbertCarreras replied:
fixed

christianvuerings commented on docs/src/components/Example.js:
```diff
@@ -33,24 +34,59 @@ export default Demo;`;
+    const imports = [];
+    if (baseComponents.length > 0) {
+      imports.push(
+        `import "../node_modules/gestalt/dist/gestalt.css";\nimport { ${baseComponents.join(
```
Let's import all the CSS first (the Gestalt one should always be loaded)

AlbertCarreras replied:
fixed

AlbertCarreras replied:
![image](https://user-images.githubusercontent.com/10593890/84557580-58767a00-ace1-11ea-9c3d-519a378e5300.png)

christianvuerings commented on docs/src/components/Example.js:
```diff
@@ -33,24 +34,59 @@ export default Demo;`;
         content: {
           title,
           // description: demoConfig.description,
-          dependencies: {
```
@AlbertCarreras can we do this instead:

```
dependencies: {
  react: 'latest',
  'react-dom': 'latest',
  ...(baseComponents.length > 0 ? { gestalt: 'latest' } : {})
  ...(additionalComponents.includes('DatePicker') ? { 'gestalt-datepicker' : 'latest' } : {})
}
```

AlbertCarreras replied:
fixed

christianvuerings commented on flow-typed/npm/react-datepicker_v2.x.x.js:
```diff
@@ -0,0 +1,170 @@
+
+declare module 'react-datepicker' {
+
+  declare type LocaleData = {
```
@AlbertCarreras could we add a link to https://github.com/date-fns/date-fns/blob/81ab18785146405ca2ae28710cdfbb13a294ec50/src/locale/af/index.js.flow

AlbertCarreras replied:
fixed

christianvuerings commented on packages/gestalt-datepicker/package.json:
```diff
@@ -0,0 +1,52 @@
+{
+  "name": "gestalt-datepicker",
+  "version": "2.2.1",
```
@AlbertCarreras let's make sure when we merge this that they are in sync.

christianvuerings commented on packages/gestalt-datepicker/src/DatePickerTextField.js:
```diff
@@ -0,0 +1,115 @@
+              value={value}
+            />
+          </Box>
+          <Box position="absolute" right={!isRTL} left={isRTL}>
```
@AlbertCarreras let's remove the isRTL prop

AlbertCarreras replied:
fixed

AlbertCarreras commented:
comments from ashley:

- [x] flexible disabled past dates: Added example http://localhost:3000/gestalt#/DatePicker 
- [x] iconButton instead of icon in month navigation: I did not added IconButton to prevent need buttons, but reproduced exact behaviour in existing button
- [ ] UX autofocus to selectEnd date picker after selecting selectStart date

- Improvements: 

1. removed duplicated accessibility labels where Labels provided accessibility or where react-datepicker provided the same accessibility
2. moved helper/label to built-in ones on TextField
3. added logic for calendar icon absolute positioning with label/helperText

ashleyseto commented:
Thanks for making the requested changes! This is approved from the design side.
